Rebase no load error on macroexpand

bramtayl opened the pull request.

bramtayl committed:
Rebase no load error on macroexpand.

Co-authored-by: Chris Foster <chris42f@gmail.com>

bramtayl commented:
Credit for the deprecation code goes to @c42f

c42f commented:
XRef: this is a continuation of: https://github.com/JuliaLang/julia/pull/37857

Note that, generally speaking, it's better to force push your WIP branch when doing a rebase, rather than opening a new PR. This makes it easier for reviewers.

> Credit for the deprecation code goes to @c42f

Thanks. By the way, the "git standard" way is to add multiple authors is to add a `Co-authored-by` line to the git commit. See 9572202251a16037d46430f673f0ccffdd8ec651 as an example of how to do this.

bramtayl commented:
Oh, ok! Is it possible to edit the existing commit to include `Co-authored-by`? Clearly I don't have much git-fu.

bramtayl commented:
Pretty sure the failure here is unrelated (maybe there's profile deadlocking on Arch?)

c42f commented:
> Is it possible to edit the existing commit

Sure - use `git commit --amend` on your branch and edit the message, then `git push -f` to force push the edited history to your branch, which will update the PR.

bramtayl commented:
Thanks for the help! Hopefully that did it

bramtayl commented:
Bump?

JeffBezanson commented:
Could use a news entry.

bramtayl committed:
add NEWS

bramtayl commented:
Ok! How's that look?

c42f commented on stdlib/Test/test/runtests.jl:
```diff
@@ -1039,3 +1039,18 @@ end
+    @test_throws LoadError @macroexpand @test_macro_throw_1
+    # Expected LoadError instances are unwrapped as necessary
+    @test_throws LoadError("file", 111, ErrorException("Real error")) @macroexpand @test_macro_throw_1
+    @test_throws LoadError("file", 111, ErrorException("Real error")) @macroexpand @test_macro_throw_2
```
As noted above, I think we need to add something like

```suggestion
    @test_throws LoadError("file", 111, ErrorException("Real error")) @macroexpand @test_macro_throw_2
    @test_throws LoadError macroexpand(@__MODULE__, :(@test_macro_throw_1)) # deprecate form
    @test_throws ErrorException macroexpand(@__MODULE__, :(@test_macro_throw_1))
```

c42f reviewed (commented):
Looking close, I added a couple more suggestions for backward compat.

bramtayl commented on stdlib/Test/src/Test.jl:
```diff
@@ -600,16 +600,28 @@ macro test_throws(extype, ex)
+        from_macroexpand =
+            orig_expr isa Expr &&
+            orig_expr.head === :macrocall &&
+            orig_expr.args[1] in MACROEXPAND_LIKE
```
Thanks for the review, and sorry it took me a while to get back to this...I'm a bit conflicted about where to go from here. Both `@macroexpand` and `macroexpand` could be called somewhere in the body of a complicated expression, or even from inside a separate function. Automatically passing `LoadError` without any further warning seems too dangerous though. In other words, I'm not sure it's possible to have a neat deprecation here?

c42f replied:
Well, to be honest I prefer the original version of this code from https://github.com/JuliaLang/julia/blob/3ee9d8f813dde0339a598524a484d4d822498101/stdlib/Test/src/Test.jl#L592

In that case I didn't try to detect where the `LoadError` came from, so the heuristic was very simple. Basically

"If within `@test_throws` and the user code specified `LoadError`, then it's ok for the actual thrown exception not to be a load error"

This is just to prevent a fair bit of spurious test breakage in actual practical cases where I know packages test their macro error messages using `@test_throws`.

Certainly any attempt at removing `LoadError` will be a minor breaking change for some type of end user code. But if we're a little over-permissive in allowing error types thrown within `@test_throws` at least the common cases will continue working and old package tests won't be spuriously broken in practice.

c42f replied:
But I'm not sure I fully understand your comment. I'm suggesting that being more permissive is ok here (I think it's ok just to remove the `from_macroexpand` check).

bramtayl replied:
Well, I guess I don't care too much either way, so happy to just remove the `from_macroexpand` check. An example where that might be a bit dangerous: perhaps a user could get confused between `LoadError` and `ArgumentError`, accidentally write `@test_throws LoadError something_that_throws_an_argument_error`, and continue on merrily when tests pass. Then, a bug in the user's code throws some other type of error, like `KeyError`. The bug wouldn't get caught.

c42f replied:
Fair comment. My thinking was that it's useful to provide a low-friction way to keep historical code working on both older Julia versions and the current one. Preventing users from writing new bugs is also a worthwhile aim.

If we don't detect calls to `macroexpand` (or just be generally more permissive as in the original PR) I know that various package tests will break in a spurious way. Examples can be found here https://github.com/search?q=test_throws+LoadError+macroexpand&type=code

One specific example:
https://github.com/GiggleLiu/NiLangCore.jl/blob/eaf26c9cde1fa30b37ed59d345dacda19eec77ff/test/compiler.jl#L104

bramtayl replied:
New bugs seem more serious that friction to me, but I can see how reducing both would be helpful. Maybe we could throw a a warning for `@test_throws [unqualified]LoadError [@]macroexpand[1] stuff`? Something like `@warn "macroexpands no longer throw a LoadError. Test passed without checking the error type!"`

c42f replied:
It depends on your perspective ;-) If you happen to be supporting a big codebase or a widely used library which is compatible across Julia versions, it's quite unfortunate to have changes to `Base` break your tests for all old versions.

> Maybe we could throw a a warning

Right, let's just call `Base.depwarn()`. That way, old versions of library tests will continue to work and the deprecation should be actively visible. In that case it also seems safe to remove the `from_macroexpand` check.

bramtayl committed:
revisions

bramtayl committed:
Merge branch 'master' into bramtayl_no_loaderror_2

bramtayl commented:
Ok, I've added a deprecation warning to undecorated `@test_throws LoadError [@]macroexpand[1]` (not 100% I did it right), removed the `from_macroexpand` check for decorated `LoadError`s, and fixed `@test_macro_throw_2` (and rebased via GitHub)

c42f commented on stdlib/Test/src/Test.jl:
```diff
@@ -601,16 +601,33 @@ macro test_throws(extype, ex)
-            success = isa(exc, extype)
+            success =
+                if from_macroexpand && extype == LoadError && exc isa Exception
+                    Base.depwarn("macroexpands no longer throw a LoadError. @test_throws passed without checking the error type!", Symbol("@test_throws LoadError [@]macroexpand[1]"))
```
`@test_deprecated` expects the message passed to `depwarn()` to contain the string "deprecated". This is a bit unfortunate in this case, but we can reword a little without it becoming too awkward.

Also, I think need to use :do_test_throws as the second argument here so that the caller of the deprecation will be found correctly.

```suggestion
                    Base.depwarn("macroexpand no longer throw a LoadError so `@test_throws LoadError ...` is deprecated and passed without checking the error type!", :do_test_throws)
```

c42f reviewed (commented):
Nice, we're very close here. `depwarn()` is subtle to use correctly.

There's some test errors here to investigate. I'm not sure whether we still run all tests with `--depwarn=error` by default, but that might be the cause.  In that case, these deprecation tests might need to be treated the same as `test/deprecation_exec.jl` (see `test/misc.jl` for how that is run)

bramtayl committed:
Fix depwarn

Co-authored-by: Chris Foster <chris42f@gmail.com>

bramtayl committed:
Fix news

bramtayl committed:
fix tests

bramtayl committed:
Merge branch 'master' into bramtayl_no_loaderror_2

bramtayl commented:
Look at all that green!

StefanKarpinski commented:
Bump. @JeffBezanson?

vtjnash commented:
> Credit for the deprecation code goes to @c42f

My apologies to @c42f that I unintentionally removed this mention when trying to fix up the squash message

bramtayl commented:
Wooo!!!
